exhaust.py

```diff
@@ -0,0 +1,176 @@
+#!/usr/bin/env python3
+
+import json
+import os
+import glob
+import multiprocessing as mp
+import progressbar as pb
+import time
+from contextlib import redirect_stdout
+
+from fpgaperf import *
+import sow
+
+MANDATORY_CONSTRAINTS = {
+    "vivado": ["xdc"],
+    "vpr": ["pcf"],
+}
+
+# to find data files
+root_dir = os.path.dirname(os.path.abspath(__file__))
+src_dir = root_dir + '/src'
+
+
+def get_families(project):
+    return matching_pattern(
+        os.path.join(src_dir, project, '*/'), '.*\/(.*)\/$'
+    )
+
+
+def get_devices(project, family):
+    return matching_pattern(
+        os.path.join(src_dir, project, family, '*/'),
+        '.*\/([^/_]*)(?:_?)(?:[^/_]*)\/$'
+    )
+
+
+def get_packages(project, family, device):
+    return matching_pattern(
+        os.path.join(src_dir, project, family, "{}*/".format(device)),
+        '.*\/(?:[^/_]*' + device + ')(?:_?)([^/_]*)\/$'
+    )
+
+
+def get_boards(project, family, device, package):
+    boards = glob.glob(
+        os.path.join(
+            src_dir, project, family, "{}_{}/*".format(device, package)
+        )
+    )
+    return sorted([board.split('/')[-1] for board in boards])
+
+
+def get_reports(out_prefix):
+    return matching_pattern(
+        root_dir + '/' + out_prefix + '/*/meta.json', '(.*)'
+    )
+
+
+def user_selected(option):
+    return [option] if option else None
+
+
+def iter_options(args):
+    for project in user_selected(args.project) or get_projects():
+        for toolchain in user_selected(args.toolchain) or get_toolchains():
+            for family in user_selected(args.family) or get_families(project):
+                for device in user_selected(args.device) or set(get_devices(
+                        project, family)):
+                    for package in user_selected(args.package) or get_packages(
+                            project, family, device):
+                        for board in user_selected(args.board) or get_boards(
+                                project, family, device, package):
```

Comment:
all those nested loops look really weird. Please use `itertools.product` instead

Reply:
Here we build the various lists iteratively, I don't think product can be used in this case. This is so far the cleaner way to iterate through all the possible combinations, based on the directory structure. I added some comments to make things clearer.

Reply:
@kgugala I have restructured once again the file structures. This because the whole way of addressing constraints files within subdirectories resulted to be counterintuitive as well as tricky to handle.

Now all the constraints file are in the `constr` directory of each project, under the relative toolchain.

Example: `oneblink/constr/vpr/xc7_a35t_csg324-1_arty.pcf`

This may result in a lot of constraints when more parts are added, but it is far easier to handle.